### PR TITLE
Implement sparsemax for attention

### PR DESCRIPTION
**Pull request by bpopeters**
This PR implements sparsemax [(Martins & Astudillo, 2016)](https://arxiv.org/pdf/1602.02068.pdf), a softmax alternative which outputs sparse probabilities but is differentiable nonetheless, and allows it to be used in an attention mechanism. It shouldn't be too hard to extends this to additionally allow constrained (soft|sparse)max attention [(Malaviya et al., 2018)](https://arxiv.org/pdf/1805.08241.pdf).

I also changed the nn.Module softmax and tanh in GlobalAttention into calls to the functions in torch.nn.functional. Using nn.Module gives GlobalAttention superfluous attributes.

### CODE REVIEW

**Comment by rbommasanimozilla** [comment]
Should also be easy to add on implementations for fusedmax + oscarmax (Niculae and Blondel, NIPS 17 - https://arxiv.org/abs/1705.07704). I think this is also very good for adding to the copying mechanism (i.e. if the copying distribution is not the same as the attention distribution) as we can see that this has some very interesting, research-worth implications on such models.

**Comment by bpopeters** [comment]
Good idea! I set the copy attention up to use the same function as the standard attention, but in the long run more flexibility is better.

**Comment by rbommasanimozilla** [comment]
I haven't entirely read through your PR, but seeing as there is already an existing way to have 2 separate distributions, this work can probably be streamlined with how that was done. Also, this is super cool and was something I was actually very erratically working on =).

**Comment by bpopeters** [comment]
Update: I've added onmt/modules/sparsemax_loss.py, implementing the sparsemax loss as described in (Martins & Astudillo, 2016). The motivation for this is slightly different -- you don't need a loss function for attention, but you do if you want to use sparsemax for generating your model outputs. This is mostly unexplored in the literature I've been able to find, but anecdotally seems to work quite well on my favorite toy problem, grapheme-to-phoneme conversion.

**Comment by kaihuchen** [comment]
Would love to try this out with my speech-to-pinyin (Chinese) experiment. I have been using [PR759](https://github.com/OpenNMT/OpenNMT-py/pull/759)  (sharding + Baidu engine, from @da03 ) for my experiments, and I wonder after the merge whether this sparsemax feature will work with PR759?
